ESPixelStick/src/output/OutputUart.hpp

```diff
@@ -171,49 +151,66 @@ inline void IRAM_ATTR StartNewDataFrame()
     uint32_t RxIsr = 0;
     uint32_t ErrorIsr = 0;
     uint32_t IsrIsNotForUs = 0;
-    uint32_t IntensityBytesSent = 0;
+    uint32_t IntensityValuesSent = 0;
     uint32_t IntensityBitsSent = 0;
-    uint32_t IntensityBytesSentLastFrame = 0;
+    uint32_t IntensityValuesSentLastFrame = 0;
     uint32_t IntensityBitsSentLastFrame = 0;
     uint32_t IncompleteFrame = 0;
     uint32_t IncompleteFrameLastFrame = 0;
     uint32_t EnqueueCounter = 0;
+    uint32_t BreakIsrCounter = 0;
+    uint32_t IdleIsrCounter = 0;
+    uint32_t TimerIsrCounter = 0;
+    uint32_t TimerIsrNoDataToSend = 0;
+    uint32_t TimerIsrSendData = 0;
+    uint32_t FiFoNotEmpty = 0;
+    uint32_t FiFoEmpty = 0;
+
 #endif // def USE_UART_DEBUG_COUNTERS
 
-#define DisableUartInterrupts CLEAR_PERI_REG_MASK(UART_INT_ENA(OutputUartConfig.UartId), UART_TXFIFO_EMPTY_INT_ENA);
-#define EnableUartInterrupts  SET_PERI_REG_MASK  (UART_INT_ENA(OutputUartConfig.UartId), UART_TXFIFO_EMPTY_INT_ENA);
+#ifndef UART_TX_BRK_DONE_INT_ENA
+#   define UART_TX_BRK_DONE_INT_ENA 0
+#endif // ndef | UART_TX_BRK_DONE_INT_ENA
+
+#ifndef UART_INTR_MASK
+#   if defined(ARDUINO_ARCH_ESP32)
+#       define UART_INTR_MASK uint32_t((1 << 18) - 1)
+#   elif defined(ARDUINO_ARCH_ESP8266)
+#       define UART_INTR_MASK uint32_t((1 << 8) - 1)
+#   endif //  defined(ARDUINO_ARCH_ESP8266)
+#endif // UART_INTR_MASK
+
+#if defined(ARDUINO_ARCH_ESP32)
+#   define WeNeedAtimer false
+#else
+#define WeNeedAtimer (OutputUartConfig.NumBreakBitsAfterIntensityData || OutputUartConfig.NumExtendedStartBits)
```

Comment:
Consider making this at least an inline function (taking as input the `OutputUartConfig`).

With the right GCC attributes, it will be treated identically to the `#define`, except you'll get scope and type safety (and all the other benefits of avoiding the preprocessor).

For example, it might look something close to the following:
```C++
inline bool WeNeedAtimer(OutputUartConfig_t& config) __attribute__ (( gnu_inline, const )) {
#if defined(ARDUINO_ARCH_ESP32)
    return false;
#else
    return (config.NumBreakBitsAfterIntensityData || config.NumExtendedStartBits);
#endif
}
```

Reply:
Pure Opinion: For some of us, the preprocessor is our friend. This latest disfavor with the preprocessor is largely due to the abuse of the preprocessor abilities to mimic template like functionality at a time that the C++ pushers wanted people to use the new functionality.

Macro has been converted.

Reply:
I have just converted this to a function in the .hpp file. You will notice that I avoid creating functions in the include files because (while we are not using it) it is my opinion that every function needs to be unit tested and functions defined in headers are notoriously difficult to test. I convinced myself that one line actions (aka macro) are an acceptable exception but a full function definition still feels like it needs a unit test and should be defined in a cpp file. I like the C# model in which header files are a thing of the past and this quandary goes away.